Add troubleshooting of node maintenance mode (#619)
harvester/docs: docs/troubleshooting/host.md

This options is available from Harvester v1.3.1 with the embedded Longhorn v1.6.0.

From Harvester v1.4.0, this option is set to `false` by default.

#### Workaround 2: Manually Detach the Volume

Detach the volume from the [Embedded Longhorn UI](./harvester.md#access-embedded-rancher-and-longhorn-dashboards).

![detached-volume.png](/img/v1.3/troubleshooting/detached-volume.png)

After that, the node will enter maintenance mode successfully.

![node-enter-maintenance-mode.png](/img/v1.3/troubleshooting/node-enter-maintenance-mode.png)

### The Single-replica Volume

Harvester supports to define the customized `StorageClass`, the [Number of Replicas](../advanced/storageclass.md#number-of-replicas) can even be 1 in some scenarios.

When such a volume was ever attached to a certain host by CSI driver or other ways, the last and only replica stays on this node even after the volume is detached from the node.

This can be checked from the CRD object `Volume`.

```
- apiVersion: longhorn.io/v1beta2
  kind: Volume
...
  spec:
...
    numberOfReplicas: 1  // the replica number
...
  status:
...
    ownerID: nodeName
...
    state: attached
```

> #### Workaround 1: Set Longhorn option `Node Drain Policy`

Reviewer: NIT: This appears to be the only workaround.

```suggestion
#### Workaround: Set Longhorn option `Node Drain Policy`
```

Author: Updated. Initially I plan to find more workarounds :)

> #### Workaround: Set Longhorn option `Node Drain Policy`
> 
> The Longhorn [Node Drain Policy](https://longhorn.io/docs/1.6.0/references/settings/#node-drain-policy) defaults to `block-if-contains-last-replica`. Longhorn will block the drain when the node contains the last healthy replica of a volume.
> 
> Set this option to `allow-if-replica-is-stopped` from the [Embedded Longhorn UI](./harvester.md#access-embedded-rancher-and-longhorn-dashboards) will solve this issue.
> 
> :::note important
> 
> If you plan to remove this node after it enters the maintenance mode, it is recommended to backup those single-replica volumes or redeploy the related workloads to other node in advance to get the volume scheduled to other node. Otherwise, those volumes can't be rebuilt or restored from other nodes after this node is removed.
> 
> :::
> 
> From Harvester v1.4.0, this option is set to `allow-if-replica-is-stopped` by default.

Reviewer 2: ```suggestion
#### Workaround: Set `Node Drain Policy`

The Longhorn [Node Drain Policy](https://longhorn.io/docs/1.6.0/references/settings/#node-drain-policy) is set to `block-if-contains-last-replica` by default. This option forces Longhorn to block node draining when the node contains the last healthy replica of a volume.

To address the issue, change the value to `allow-if-replica-is-stopped` using the [embedded Longhorn UI](./harvester.md#access-embedded-rancher-and-longhorn-dashboards).

:::info important

If you plan to remove the node after Maintenance Mode is enabled, back up single-replica volumes or redeploy the related workloads to other nodes in advance so that the volumes are scheduled to other nodes.

:::

Starting with Harvester v1.4.0, the Node Drain Policy is set to `allow-if-replica-is-stopped` by default.
```

:::

#### Workaround 1: Set `Detach Manually Attached Volumes When Cordoned` to `True`

The Longhorn setting [Detach Manually Attached Volumes When Cordoned](https://longhorn.io/docs/1.6.0/references/settings/#detach-manually-attached-volumes-when-cordoned) blocks node draining when there are volumes manually attached to the node.

The default value of this setting depends on the embedded Longhorn version:

| Harvester version | Embedded Longhorn version | Default value |
| --- | --- | --- |
| v1.3.1 | v1.6.0 | `true` |
| v1.4.0 | v1.7.0 | `false` |

Set this option to `true` from the [embedded Longhorn UI](./harvester.md#access-embedded-rancher-and-longhorn-dashboards).

#### Workaround 2: Manually Detach the Volume

Detach the volume using the [embedded Longhorn UI](./harvester.md#access-embedded-rancher-and-longhorn-dashboards).

![detached-volume.png](/img/v1.3/troubleshooting/detached-volume.png)

Once the volume is detached, you can successfully enable Maintenance Mode on the node.

![node-enter-maintenance-mode.png](/img/v1.3/troubleshooting/node-enter-maintenance-mode.png)

### Single-Replica Volumes

> Harvester allows you to create custom StorageClasses that describe how Longhorn must provision volumes. If necessary, you can create a StorageClass with the [Number of Replicas](../advanced/storageclass.md#number-of-replicas) parameter set to `1`.

Reviewer 3: ```suggestion
Harvester allows you to create custom `StorageClasses` that describe how Longhorn must provision volumes. If necessary, you can create a StorageClass with the [Number of Replicas](../advanced/storageclass.md#number-of-replicas) parameter set to `1`.
```

> When a volume is created using such a StorageClass and is attached to a node using the CSI driver or other methods, the lone replica stays on that node even after the volume is detached.

Reviewer 3: Use inline-code as previously already done to keep consistency across the document.
```suggestion
When a volume is created using such a `StorageClass` and is attached to a node using the CSI driver or other methods, the lone replica stays on that node even after the volume is detached.
```

You can check this using the CRD object `Volume`.

```
- apiVersion: longhorn.io/v1beta2
  kind: Volume
...
  spec:
...
    numberOfReplicas: 1  // the replica number
...
  status:
...
    ownerID: nodeName
...
    state: attached
```

#### Workaround: Set `Node Drain Policy`

The Longhorn [Node Drain Policy](https://longhorn.io/docs/1.6.0/references/settings/#node-drain-policy) is set to `block-if-contains-last-replica` by default. This option forces Longhorn to block node draining when the node contains the last healthy replica of a volume.

To address the issue, change the value to `allow-if-replica-is-stopped` using the [embedded Longhorn UI](./harvester.md#access-embedded-rancher-and-longhorn-dashboards).

:::info important

> If you plan to remove the node after Maintenance Mode is enabled, back up single-replica volumes or redeploy the related workloads to other nodes in advance so that the volumes are scheduled to other nodes.

Reviewer 3: Does it make sense to highlight the `Mainenance Mode` using inline-code across the whole document? E.g. below `Node Drain Policy` is highlighted as well.

@jillian-maroket Is there a rule when to use inline-code?
